File: demos/ik/end_effector_tasks.py
## This file contains apis to create different end effector tasks
## for the inverse kinematics problem
## Author : Avadesh Meduri
## Date : 24/02/2021

import numpy as np
import pinocchio as pin
import crocoddyl

class EndEffectorTasks:

    def __init__(self):
        ## This class contains different tasks for swing foot trajectory
        ## This is inherited by the inversekinematics class
        ## Note :
            ## Paramteres such as self.dt etc.. are defined in the inverse kinematics
            ## class
        pass

    def add_trajectory_tracking_task(self, fid, st, et, traj, wt, cost_name):
        """
        This function creates a end effector tracking task in the ddp
        Input:
            fid : the frame id of the frame that should track the trajectory
            st : start time (sec)
            et : end time (sec)
            traj : the trajectory to be tracked (must have shape ((en - sn)/dt, 3))
            wt : weight of the cost
            cost_name : name of the cost
        """
        sn, en = int(st/self.dt), int(et/self.dt)
        for i in range(sn, en):
            Mref = crocoddyl.FrameTranslation(fid, traj[i - sn])
            goalTrackingCost = crocoddyl.CostModelFrameTranslation(self.state, Mref)
            self.rcost_model_arr[i].addCost(cost_name, goalTrackingCost, wt)
            
    def add_velocity_tracking_cost(self, fid, st, et, traj, wt, cost_name):
        """
        This function creates a task for the end effector to track a desired velocity
        Input:
            fid : the frame id of the frame that should track the trajectory
            st : start time (sec)
            et : end time (sec)
            traj : the velocity trajectory to be tracked (must have shape ((en - sn)/dt, 6))
            wt : weight of the cost
            cost_name : name of the cost
        """
        sn, en = int(st/self.dt), int(et/self.dt)
        for i in range(sn, en):
            Vref = crocoddyl.FrameMotion(fid, pin.Motion(traj[i - sn]))
            velTrackingCost = crocoddyl.CostModelFrameVelocity(self.state, Vref)
            self.rcost_model_arr[i].addCost(cost_name, velTrackingCost, wt)

File: src/ik/end_effector_tasks.cpp
// This file contains end effector tracking tasks

#include "ik/inverse_kinematics.hpp"


namespace ik{

    void InverseKinematics::add_position_tracking_task(
                    pinocchio::FrameIndex fid, int sn, int en, 
                    Eigen::MatrixXd traj, double wt, std::string cost_name){

        for (unsigned i = sn; i < en; ++i){
            boost::shared_ptr<crocoddyl::CostModelAbstract> goal_tracking_cost =
                boost::make_shared<crocoddyl::CostModelResidual>(state_, 
                boost::make_shared<crocoddyl::ResidualModelFrameTranslation>(state_, fid, traj));
            rcost_arr_[i].get()->addCost(cost_name + std::to_string(i), goal_tracking_cost, wt);
        }
    };

    void InverseKinematics::add_position_tracking_task_single(pinocchio::FrameIndex fid, Eigen::MatrixXd traj,
        double wt, std::string cost_name, int time_step){

            boost::shared_ptr<crocoddyl::CostModelAbstract> goal_tracking_cost =
                    boost::make_shared<crocoddyl::CostModelResidual>(state_, 
                    boost::make_shared<crocoddyl::ResidualModelFrameTranslation>(state_, fid, traj));
            rcost_arr_[time_step].get()->addCost(cost_name, goal_tracking_cost, wt);
    };

    void InverseKinematics::add_terminal_position_tracking_task(
                    pinocchio::FrameIndex fid, Eigen::MatrixXd traj, 
                                double wt, std::string cost_name){

        boost::shared_ptr<crocoddyl::CostModelAbstract> goal_tracking_cost =
                    boost::make_shared<crocoddyl::CostModelResidual>(state_, 
                    boost::make_shared<crocoddyl::ResidualModelFrameTranslation>(state_, fid, traj));
        tcost_model_.get()->addCost(cost_name, goal_tracking_cost, wt);
    };

    void InverseKinematics::add_position_tracking_task_single(pinocchio::FrameIndex fid, Eigen::MatrixXd traj,
        Eigen::VectorXd weight, std::string cost_name, int time_step){
            boost::shared_ptr<crocoddyl::ActivationModelAbstract> goal_activation =
                                        boost::make_shared<crocoddyl::ActivationModelWeightedQuad>(weight);
            boost::shared_ptr<crocoddyl::CostModelAbstract> goal_tracking_cost =
                    boost::make_shared<crocoddyl::CostModelResidual>(state_, 
                    goal_activation,
                    boost::make_shared<crocoddyl::ResidualModelFrameTranslation>(state_, fid, traj));
            rcost_arr_[time_step].get()->addCost(cost_name, goal_tracking_cost, 1.0);
    };

    void InverseKinematics::add_terminal_position_tracking_task(
                    pinocchio::FrameIndex fid, Eigen::MatrixXd traj, 
                                Eigen::VectorXd weight, std::string cost_name){
                            
        boost::shared_ptr<crocoddyl::ActivationModelAbstract> goal_activation =
                                        boost::make_shared<crocoddyl::ActivationModelWeightedQuad>(weight);
        boost::shared_ptr<crocoddyl::CostModelAbstract> goal_tracking_cost =
                boost::make_shared<crocoddyl::CostModelResidual>(state_, 
                goal_activation,
                boost::make_shared<crocoddyl::ResidualModelFrameTranslation>(state_, fid, traj));
        tcost_model_.get()->addCost(cost_name, goal_tracking_cost, 1.0);
    };

    void InverseKinematics::add_velocity_tracking_task(
                    pinocchio::FrameIndex fid, int st, int et, 
                    Eigen::MatrixXd traj, double wt, std::string cost_name){

        std::cout << "function not implemented" << std::endl;
        // for (unsigned i = sn; i < en; ++i){
        //     crocoddyl::FrameMotion Mref(fid, pinocchio::Motion(traj.row(i - sn)));
        //     boost::shared_ptr<crocoddyl::CostModelAbstract> vel_tracking_cost =
        //         boost::make_shared<crocoddyl::CostModelResidual>(
        //             state_,
        //             boost::make_shared<crocoddyl::ResidualModelFrameVelocity>(
        //                 state_, fid, pinocchio::Motion(traj.row(i - sn))));

        //     // boost::shared_ptr<crocoddyl::CostModelAbstract> vel_tracking_cost = 
        //     //     boost::make_shared<crocoddyl::CostModelFrameVelocity>(state_, Mref);   

        //     rcost_arr_[i].get()->addCost(cost_name, vel_tracking_cost, wt);
        // };
    };



}